propulsion-hotel/Reactor/Handler.fs

```diff
@@ -0,0 +1,62 @@
+module Reactor.Handler
+
+open Infrastructure
+open Propulsion.Internal
+
+type Outcome = GroupCheckoutProcess.Outcome
+
+/// Gathers stats based on the outcome of each Span processed, periodically including them in the Sink summaries
+type Stats(log, statsInterval, stateInterval, ?logExternalStats) =
+    inherit Propulsion.Streams.Stats<Outcome>(log, statsInterval, stateInterval)
+
+    let mutable completed, ignored, failed, succeeded = 0, 0, 0, 0
+
+    override _.DumpStats() =
+        base.DumpStats()
+        if completed <> 0 || ignored <> 0 || failed <> 0 || succeeded <> 0 then
+            log.Information(" Merged {completed} failed {failed} succeeded {succeeded} Ignored {ignored} ", completed, failed, succeeded, ignored)
+            completed <- 0; failed <- 0; succeeded <- 0; ignored <- 0
+        match logExternalStats with None -> () | Some f -> let logWithoutContext = Serilog.Log.Logger in f logWithoutContext
+
+    override _.HandleOk res = res |> function
+        | Outcome.Merged (ok, denied)-> completed <- completed + 1; succeeded <- succeeded + ok; failed <- failed + denied
+        | Outcome.Noop -> ignored <- ignored + 1
+    override _.Classify(exn) =
+        match exn with
+        | Equinox.DynamoStore.Exceptions.ProvisionedThroughputExceeded -> Propulsion.Streams.OutcomeKind.RateLimited
+        | x -> base.Classify x
+    override _.HandleExn(log, exn) =
+        log.Information(exn, "Unhandled")
+
+open Domain
+
+let private isReactionStream = function
+    | GroupCheckout.Category -> true
+    | _ -> false
+
+let private handle (processor : GroupCheckoutProcess.Service) stream = async {
+    match stream with
+    | GroupCheckout.StreamName groupCheckoutId ->
+        let! outcome, ver' = processor.React(groupCheckoutId)
+        return struct (Propulsion.Streams.SpanResult.OverrideWritePosition ver', outcome)
```

Comment:
I think this deserves a comment. Something along these lines.

```suggestion
        // Checkpointing happens asynchronously. When the process is started up we might reprocess
        // already handled messages. The service returns the version of the process manager stream
        // so we can use that to fast-forward our processing and avoid re-triggering actions unnecessarily
        return struct (Propulsion.Streams.SpanResult.OverrideWritePosition ver', outcome)
```

Reply:
An equally important part of this is that event 9 might trigger the handler. Then, when it reads, it might only see up to version 8. Using OverrideWritePosition (maybe that should be named something different like `ProcessedUpToVersion`) means that event 9 stays pending and it will keep being retried until the handler deems the achieved version to be beyond the event.

Will try to work that into the comment (I had a comment on the calling piece that was supposed to imply this, but on reflection it's pretty damn cryptic!)

Reply:
(much more thorough comments added, would appreciate a scan of them)